spark/src/test/scala/org/apache/spark/sql/delta/DeltaAlterTableReplaceTests.scala

```diff
@@ -0,0 +1,745 @@
+      assertEqual(deltaLog.snapshot.schema, new StructType()
+        .add("v1", "integer")
+        .add("v2", "string")
+        .add("v3", "long")
+        .add("s", new StructType()
+          .add("v1", "integer").add("v2", "string").add("v3", "long"))
+        .add("a", ArrayType(new StructType()
+          .add("v1", "integer").add("v2", "string").add("v3", "long")))
+        .add("m", MapType(
+          new StructType().add("v1", "integer").add("v2", "string").add("v3", "long"),
+          new StructType().add("v1", "integer").add("v2", "string").add("v3", "long"))))
+
+      implicit val ordering = Ordering.by[
+        (Int, String, Option[Long],
+          (Int, String, Option[Long]),
+          Seq[(Int, String, Option[Long])],
+          Map[(Int, String, Option[Long]), (Int, String, Option[Long])]), Int] {
+        case (v1, _, _, _, _, _) => v1
+      }
+      checkDatasetUnorderly(
+        spark.table(tableName).as[
+          (Int, String, Option[Long],
+            (Int, String, Option[Long]),
+            Seq[(Int, String, Option[Long])],
+            Map[(Int, String, Option[Long]), (Int, String, Option[Long])])],
+        (1, "a", None, (1, "a", None),
+          Seq((1, "a", None)), Map((1, "a", Option.empty[Long]) -> ((1, "a", None)))),
+        (2, "b", None, (2, "b", None),
+          Seq((2, "b", None)), Map((2, "b", Option.empty[Long]) -> ((2, "b", None)))))
+    }
+  }
+
+  ddlTest("REPLACE COLUMNS - special column names") {
+    val df = Seq((1, "a"), (2, "b")).toDF("x.x", "y.y")
+      .withColumn("s.s", struct("`x.x`", "`y.y`"))
+      .withColumn("a.a", array("`s.s`"))
+      .withColumn("m.m", map(col("`s.s`"), col("`s.s`")))
+    withDeltaTable(df) { tableName =>
+
+      sql(s"""
+             |ALTER TABLE $tableName REPLACE COLUMNS (
+             |  `m.m` MAP<STRUCT<`y.y`:string, `x.x`:int>, STRUCT<`y.y`:string, `x.x`:int>>,
+             |  `y.y` string,
+             |  `a.a` ARRAY<STRUCT<`y.y`:string, `x.x`:int>>,
+             |  `x.x` int,
+             |  `s.s` STRUCT<`y.y`:string, `x.x`:int>
+             |)""".stripMargin)
+
+      val deltaLog = getDeltaLog(tableName)
+      assertEqual(deltaLog.snapshot.schema, new StructType()
+        .add("m.m", MapType(
+          new StructType().add("y.y", "string").add("x.x", "integer"),
+          new StructType().add("y.y", "string").add("x.x", "integer")))
+        .add("y.y", "string")
+        .add("a.a", ArrayType(new StructType().add("y.y", "string").add("x.x", "integer")))
+        .add("x.x", "integer")
+        .add("s.s", new StructType().add("y.y", "string").add("x.x", "integer")))
+    }
+  }
+
+  ddlTest("REPLACE COLUMNS - drop column") {
+    // Column Mapping allows columns to be dropped
+    def checkReplace(
+                      text: String,
+                      tableName: String,
+                      columnDropped: Seq[String],
+                      messages: String*): Unit = {
+      if (!columnMappingEnabled) {
```

Comment:
nit: negated logic is error prone... any reason we can't reverse condition and the if/else blocks?